### DISCUSSION TITLE
What does `pure`, `lift`, and `sublift` mean in JAX core?

### DISCUSSION THREAD

**Discussion by leiteg** [jax-ml/jax#18270] [Q&A] [answered]
After reading the [JAX core from scratch](https://jax.readthedocs.io/en/latest/autodidax.html) and reading some parts of the JAX source code, I was confused as to what `pure`, `lift` and `sublift` methods in `core.Trace` mean? Do they simply take a JAX array and create an `ShapedArray` from them?

**Comment by mattjj** [comment, marked as answer]
Thanks for the question!

It may be tricky to answer in full detail, so instead I hope this summary answer is still useful:
* `trace.pure` is called on non-`Tracer` values, e.g. numpy.ndarrays or Python builtin number type instances, and produces a `Tracer` for the `trace` meant to represent having trivial context added (e.g. a zero tangent for `JVPTrace`, a `None` batch dim for `BatchTrace`, etc);
* `trace.lift` is similar but is always called on `Tracer` values, specifically ones that don't belong to the `trace` (e.g. if we're doing `jvp`-of-`jvp`, or `vmap`-of-`jvp`, then the inner one's `trace.lift` might be applied to a `Tracer` from the outer one), and also produces a `Tracer` for the `trace` meant to represent trivial context;
* `trace.sublift` is similar but is always called on `Tracer` values, specifically ones that _do_ belong to the same `trace`, but from outside the scope of special kinds of higher-order primitives (like `custom_jvp`).

The last one has no analogue in Autodidax (the JAX core from scratch tutorial).

Does that help?

**Reply by leiteg** [reply]
Thanks @mattjj, that helps a lot! So does that mean that the method `full_raise` of `Trace` simply takes any value and delegates it to one of `pure`, `lift`, and `sublift`?

I need some time to process this information, but I am satisfied for now, marking as answered :smiley:
